### ISSUE TITLE
HTTP method GET is not supported by this URL

### ISSUE THREAD

**Issue by softshipper** [earldouglas/xsbt-web-plugin#379]
Hi all 
I am using HTTP4S and the webapp is running on jetty. The web app file is configured as:

```
<web-app xmlns="http://java.sun.com/xml/ns/javaee"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://java.sun.com/xml/ns/javaee http://java.sun.com/xml/ns/javaee/web-app_3_0.xsd"
         version="3.0">


   <servlet>
      <servlet-name>user-svc</servlet-name>
      <servlet-class>io.databaker.UserSvcServlet</servlet-class>
      <async-supported>true</async-supported>
   </servlet>

   <servlet-mapping>
      <servlet-name>user-svc</servlet-name>
      <url-pattern>/*</url-pattern>
   </servlet-mapping>

</web-app>

```
The available URI's are:
```
object UserSvcRoutes {

  def helloWorldRoutes[F[_]: Sync](H: HelloWorld[F]): HttpRoutes[F] = {
    val dsl = new Http4sDsl[F]{}
    import dsl._
    HttpRoutes.of[F] {
      case GET -> Root =>
        Ok("Example")
      case GET -> Root / "hello" / name =>
        for {
          greeting <- H.hello(HelloWorld.Name(name))
          resp <- Ok(greeting)
        } yield resp
    }
  }

}
```
When I call `http://localhost:8080/` I've got: 
![Screenshot from 2020-05-30 08-11-11](https://user-images.githubusercontent.com/56756949/83321023-267b0780-a24d-11ea-8930-aba83bdcfc08.png)

What I am doing wrong? I have put the source code as an attachment. 

[main.zip](https://github.com/earldouglas/xsbt-web-plugin/files/4705167/main.zip)

Thanks

**Comment by earldouglas** [comment]
Can you post your sbt configuration as well?  With that, I can try to reproduce the error.

**Comment by softshipper** [comment]
Here we go
```
val Http4sVersion = "0.21.4"
val CirceVersion = "0.13.0"
val Specs2Version = "4.9.3"
val LogbackVersion = "1.2.3"
val ServletVersion = "4.0.1"

lazy val root = (project in file("."))
  .enablePlugins(JettyPlugin)
  .settings(
    organization := "io.databaker",
    name := "user-svc",
    version := "0.0.1-SNAPSHOT",
    scalaVersion := "2.13.2",
    mainClass := Some("io.databaker.Main"),
    libraryDependencies ++= Seq(
      "org.http4s" %% "http4s-servlet" % Http4sVersion,
      "org.http4s" %% "http4s-circe" % Http4sVersion,
      "org.http4s" %% "http4s-dsl" % Http4sVersion,
      "io.circe" %% "circe-generic" % CirceVersion,
      "org.specs2" %% "specs2-core" % Specs2Version % "test",
      "ch.qos.logback" % "logback-classic" % LogbackVersion,
      "javax.servlet" % "javax.servlet-api" % ServletVersion % "provided"
    ),
    addCompilerPlugin("org.typelevel" %% "kind-projector" % "0.10.3"),
    addCompilerPlugin("com.olegpy" %% "better-monadic-for" % "0.3.1"),
/*    // disable .jar publishing
    Compile / packageBin / publishArtifact := false,

    // create an Artifact for publishing the .war file
    Compile / packageWar / artifact := {
      val prev: Artifact = (Compile / packageWar / artifact).value
      prev.withType("war").withExtension("war")
    },

    // add the .war file to what gets published
    addArtifact(Compile / packageWar / artifact, packageWar),*/
  )

scalacOptions ++= Seq(
  "-deprecation",
  "-encoding", "UTF-8",
  "-language:higherKinds",
  "-language:postfixOps",
  "-feature",
  "-Xfatal-warnings",
)
```
Thanks a lot

**Comment by earldouglas** [comment]
It looks like [`Http4sServlet`][1] was recently made abstract, with two concrete implementations provided by [`BlockingHttp4sServlet.scala`][2] and [`AsyncHttp4sServlet`][3].  You can get your example working by changing `UserSvcServlet` to extend either of these:

```scala
package io.databaker

import AppContextShift._
import cats.effect._
import java.util.concurrent.Executors
import org.http4s.server.DefaultServiceErrorHandler
import org.http4s.servlet.BlockingHttp4sServlet
import org.http4s.servlet.BlockingServletIo

class UserSvcServlet
  extends BlockingHttp4sServlet[IO](
    service = UserSvcServer.start,
    servletIo = BlockingServletIo(4096, Blocker.liftExecutorService(Executors.newCachedThreadPool())),
    serviceErrorHandler = DefaultServiceErrorHandler
  )
```

[1]: https://github.com/http4s/http4s/blob/v0.21.4/servlet/src/main/scala/org/http4s/servlet/Http4sServlet.scala
[2]: https://github.com/http4s/http4s/blob/v0.21.4/servlet/src/main/scala/org/http4s/servlet/BlockingHttp4sServlet.scala
[3]: https://github.com/http4s/http4s/blob/v0.21.4/servlet/src/main/scala/org/http4s/servlet/AsyncHttp4sServlet.scala

**Comment by softshipper** [comment]
@earldouglas How did you figure out? I was searching for the whole morning!!! Thanks a lot man. 
Thanks for creating the wonderful plugin.

**Comment by softshipper** [comment]
@earldouglas How did you figure out? I was searching for the whole morning!!! Thanks a lot man. 
Thanks for creating the wonderful plugin.

**Comment by earldouglas** [comment]
Glad it's working for you!

> How did you figure out?

The error lacked any stack trace or http4s-specific information, which made me suspect a bad servlet implementation.  I checked the source code of `Http4sServlet`, and saw that it didn't have a `service` method, which is part of the servlet specification, so I knew something had changed.  Jumping up to the package level, I saw `AsyncHttp4sServlet` and `BlockingHttp4sServlet`, and figured that's what had been added.

I updated the [example http4s application][1] to use `AsyncHttp4sServlet` from 0.21.4, in case you're curious to see an async example.

[1]: https://github.com/earldouglas/xsbt-web-plugin/blob/5a60972/docs/examples/http4s/src/main/scala/HelloWorldServlet.scala
